app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/BaseFragment.kt

```diff
@@ -1,5 +1,88 @@
 package com.bottlerocketstudios.brarchitecture.ui
 
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.annotation.CallSuper
+import androidx.annotation.LayoutRes
+import androidx.databinding.DataBindingUtil
+import androidx.databinding.ViewDataBinding
 import androidx.fragment.app.Fragment
+import com.bottlerocketstudios.brarchitecture.BR
 
-abstract class BaseFragment : Fragment()
+/**
+ * Manages DataBinding setup and navigation observer setup, allows per fragment logic/customizations via [setupBinding]
+ *
+ * To create a new fragment:
+ *
+ * 1. Pass in the associated fragment ViewModel (extending [BaseViewModel]) and Binding on the [BaseFragment] subclass.
+ * 2. Override [fragmentViewModel] (using `by viewModel()`, `by sharedViewModel()`, or similar function) and [getLayoutRes] with the fragment's `R.layout.foo` resource identifier.
+ * 3. Override [setupBinding] instead of [onCreateView] or [onViewCreated].
+ */
+abstract class BaseFragment<FRAGMENT_VIEW_MODEL : BaseViewModel, BINDING : ViewDataBinding>() : Fragment() {
+
+    /** Fragment scoped Android ViewModel. Primary ViewModel to interact with for the given Fragment. */
+    protected abstract val fragmentViewModel: FRAGMENT_VIEW_MODEL
+
+    // Use private nullable binding in order to set in onCreateView and onDestroyView
+    private var _binding: BINDING? = null
+
+    /**
+     * Wraps private nullable property. Lifecycle valid from [onCreateView] to [onDestroyView].
+     * Treat similar to [requireContext] which throws if context is null, meaning you are aware of when you can call it and should use it.
+     */
+    private val binding: BINDING
+        get() = _binding!!
+
+    /**
+     * See [setupBinding] if you want to add logic post view creation.
+     *
+     * > It is recommended to **only** inflate the layout in this method and move logic that operates on the returned View to onViewCreated(View, Bundle).
+     * > Source: https://developer.android.com/reference/androidx/fragment/app/Fragment#onCreateView(android.view.LayoutInflater,%20android.view.ViewGroup,%20android.os.Bundle)
+     */
+    @CallSuper
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        _binding = DataBindingUtil.inflate(inflater, getLayoutRes(), container, false)
+        return binding.root
+    }
+
+    /** Calls [setupBinding] */
+    @CallSuper
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        setupBinding(binding)
+    }
+
+    /** Use onDestroy(binding) if need access to [binding]. Otherwise, override this function and call super. */
+    @CallSuper
+    override fun onDestroyView() {
+        super.onDestroyView()
+        onDestroyView(binding)
```

Comment:
Is it important that the binding is explicitly destoryed as so?

Reply:
@AaronEvans-BR Yes, explicitly nulling it out prevents it from unintentionally being used when it is invalid (after `onDestroyView` but before `onDestroy`) as well as satisfies LeakCanary leaks warnings. See the stackoverflow link mentioned on line 81 for more info :)